android/src/main/java/com/oblador/keychain/cipherStorage/CipherStorageKeystoreRSAECB.java

```diff
@@ -0,0 +1,339 @@
+            }
+            mDecryptParams = null;
+        }
+    }
+
+    private boolean canStartFingerprintAuthentication() {
+        return (mKeyguardManager.isKeyguardSecure() && mContext.checkSelfPermission(Manifest.permission.USE_FINGERPRINT) == PackageManager.PERMISSION_GRANTED)
+    }
+
+    private void startFingerprintAuthentication() throws Exception {
+        // If we have a previous cancellationSignal, cancel it.
+        if (mBiometricPromptCompatCancellationSignal != null) {
+            mBiometricPromptCompatCancellationSignal.cancel();
+        }
+
+        if (mActivity == null) {
+            throw new Exception("mActivity is null (make sure to call setCurrentActivity)");
+        }
+
+        mBiometricPromptCompatCancellationSignal = new CancellationSignal();
+
+        mBiometricPromptCompat = new BiometricPromptCompat.Builder(mActivity)
+            .setTitle("Authentication required")
+            .setSubtitle("Please use biometric authentication to unlock the app")
+            .build();
+
+        mBiometricPromptCompat.authenticate(mBiometricPromptCompatCancellationSignal, this);
+    }
+
+    @Override
+    public String getCipherStorageName() {
+        return CIPHER_STORAGE_NAME;
+    }
+
+    @Override
+    public boolean getCipherBiometrySupported() {
+        return true;
+    }
+
+    @Override
+    public int getMinSupportedApiLevel() {
+        return Build.VERSION_CODES.M;
+    }
+
+    @Override
+    public EncryptionResult encrypt(@NonNull String service, @NonNull String username, @NonNull String password) throws CryptoFailedException {
+        service = getDefaultServiceIfEmpty(service);
+
+        try {
+            KeyStore keyStore = getKeyStoreAndLoad();
+
+            if (!keyStore.containsAlias(service)) {
+                generateKeyAndStoreUnderAlias(service);
+            }
+
+            KeyFactory keyFactory = KeyFactory.getInstance(ENCRYPTION_ALGORITHM);
+            PublicKey publicKey = keyStore.getCertificate(service).getPublicKey();
+            KeySpec spec = new X509EncodedKeySpec(publicKey.getEncoded());
+            Key key = keyFactory.generatePublic(spec);
+
+            byte[] encryptedUsername = encryptString(key, service, username);
+            byte[] encryptedPassword = encryptString(key, service, password);
+
+            return new EncryptionResult(encryptedUsername, encryptedPassword, this);
+        } catch (NoSuchAlgorithmException | InvalidAlgorithmParameterException | NoSuchProviderException e) {
+            throw new CryptoFailedException("Could not encrypt data for service " + service, e);
+        } catch (KeyStoreException | KeyStoreAccessException e) {
+            throw new CryptoFailedException("Could not access Keystore for service " + service, e);
+        } catch (Exception e) {
+            throw new CryptoFailedException("Unknown error: " + e.getMessage(), e);
+        }
+    }
+
```

Comment:
// seems odd to provide encryption of "two things" in one call.   it may not be a username and password, for example.
// be cool to provide a getPublicKey function, so the user can pass in properly formatted public key encrypted bytes from, say, another device.

` public byte[] getPublicKey(@NonNull String service) { 
   KeyStore keyStore = getKeyStoreAndLoad();
   generateKeyAndStoreUnderAlias(service);
   KeyFactory keyFactory = KeyFactory.getInstance(ENCRYPTION_ALGORITHM);
   PublicKey publicKey = keyStore.getCertificate(service).getPublicKey();
   return publicKey.getEncoded();
}